src/workerd/api/urlpattern.c++

```diff
@@ -4,289 +4,191 @@
+    return jsg::alloc<URLPattern>(kj::mv(*result));
+  }
+
+  JSG_FAIL_REQUIRE(TypeError, "Failed to construct URLPattern");
 }
 
-kj::Maybe<URLPattern::URLPatternResult> URLPattern::exec(
+bool URLPattern::test(
     jsg::Lock& js, jsg::Optional<URLPatternInput> maybeInput, jsg::Optional<kj::String> maybeBase) {
-  auto input = kj::mv(maybeInput).orDefault(URLPattern::URLPatternInit());
-  kj::Vector<URLPattern::URLPatternInput> inputs(2);
-
-  kj::String protocol = nullptr;
-  kj::String username = nullptr;
-  kj::String password = nullptr;
-  kj::String hostname = nullptr;
-  kj::String port = nullptr;
-  kj::String pathname = nullptr;
-  kj::String search = nullptr;
-  kj::String hash = nullptr;
+  ada::url_pattern_input input = ada::url_pattern_init{};
+  std::string_view* base = nullptr;
 
-  KJ_SWITCH_ONEOF(input) {
-    KJ_CASE_ONEOF(string, kj::String) {
-      KJ_IF_SOME(url, jsg::Url::tryParse(string.asPtr(), maybeBase.map([](kj::String& s) {
-        return s.asPtr();
-      }))) {
-        auto p = url.getProtocol();
-        protocol = kj::str(p.first(p.size() - 1));
-        username = kj::str(url.getUsername());
-        password = kj::str(url.getPassword());
-        hostname = kj::str(url.getHostname());
-        port = kj::str(url.getPort());
-        pathname = kj::str(url.getPathname());
-        search = url.getSearch().size() > 0 ? kj::str(url.getSearch().slice(1)) : kj::String();
-        hash = url.getHash().size() > 0 ? kj::str(url.getHash().slice(1)) : kj::String();
-      } else {
-        return kj::none;
+  KJ_IF_SOME(i, maybeInput) {
+    KJ_SWITCH_ONEOF(i) {
+      KJ_CASE_ONEOF(string, kj::String) {
+        input = std::string_view(string.begin(), string.size());
       }
-      inputs.add(kj::mv(string));
-      KJ_IF_SOME(base, maybeBase) {
-        inputs.add(kj::mv(base));
+      KJ_CASE_ONEOF(pi, URLPattern::URLPatternInit) {
+        input = pi.toAdaType();
       }
     }
-    KJ_CASE_ONEOF(i, URLPattern::URLPatternInit) {
-      JSG_REQUIRE(
-          maybeBase == kj::none, TypeError, "A baseURL is not allowed when input is an object.");
-      inputs.add(URLPattern::URLPatternInit{
-        .protocol = i.protocol.map([](kj::String& str) { return kj::str(str); }),
-        .username = i.username.map([](kj::String& str) { return kj::str(str); }),
-        .password = i.password.map([](kj::String& str) { return kj::str(str); }),
-        .hostname = i.hostname.map([](kj::String& str) { return kj::str(str); }),
-        .port = i.port.map([](kj::String& str) { return kj::str(str); }),
-        .pathname = i.pathname.map([](kj::String& str) { return kj::str(str); }),
-        .search = i.search.map([](kj::String& str) { return kj::str(str); }),
-        .hash = i.hash.map([](kj::String& str) { return kj::str(str); }),
-        .baseURL = i.baseURL.map([](kj::String& str) { return kj::str(str); }),
-      });
+  }
+
+  KJ_IF_SOME(b, maybeBase) {
+    *base = std::string_view(b.begin(), b.size());
```

Comment:
likewise for null deref

Reply:
how come?

Reply:
base is defined as:
`std::string_view* base = nullptr;`
Then you have
`*base = ...`